### PR TITLE
Feature/boosting rules

### LINKED ISSUE #109: Add support for boosting rules
Matej 9 supports boosting rules in recommendation query.

Boosting rules will can be added similarly to sorting rules by passing instance of `Lmc\Matej\Model\Command\Boost`

```php
$recommendation = UserRecommendation::create('user-id', 'test-scenario');
$recommendation->addBoost(Boost::create('valid_to >= NOW()', 2));
```
or in batch
```php
$recommendation->addBoost([Boost::create('for_reccomendation >= NOW()', 3.5)]);
```

It will also support re-setting all boostings
```php
$recommendation->setBoosts([Boost::create('for_reccomendation >= NOW()', 3.5)]);
```

`setBoostings` will also accept empty array. In that case it will effectively remove all boostings:
```php
$recommendation->setBoosts([]);
```
`Boost` will accept exactly two arguments: `query` (`str`) and  `multiplier` (`double` > 0).

dependecy: https://github.com/lmc-eu/matej-client-php/issues/108

*There's an inconsistency between adding adding filters (which accepts only strings) and adding sorting (which accepts instances of `Sorting`) even though they are otherwise very similar.  I decided to use api consistent with sorting based on the fact that BoostingRules require more than one argument.*

### PR DESCRIPTION
**Pull request by JakubTesarek**

### CODE REVIEW

File: src/Model/Command/UserRecommendation.php [hunk 1]
```diff
@@ -239,6 +265,13 @@ protected function getCommandType(): string
         return 'user-based-recommendations';
     }
 
+    protected function serializeBoosts(): array
+    {
+        return array_map(
+            function($boost) {return $boost->jsonSerialize(); }, $this->boosts
```
Comment (OndraM) [inline src/Model/Command/UserRecommendation.php]:
```suggestion
        function (Boost $boost) {
            return $boost->jsonSerialize();
        },
        $this->boosts
```

File: src/Model/Command/UserRecommendation.php [hunk 2]
```diff
@@ -239,6 +265,13 @@ protected function getCommandType(): string
         return 'user-based-recommendations';
     }
 
+    protected function serializeBoosts(): array
```
Comment (OndraM) [inline src/Model/Command/UserRecommendation.php]:
```suggestion
    protected function getSerializedBoosts(): array
```

To make it clear from the name the method does not have sideffects.

File: tests/unit/Model/Command/BoostTest.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        $this->assertSame(
+            ['query' => 'valid_to >= NOW()', 'multiplier' => 2.1],
+            $boost->jsonSerialize()
+        );
+    }
+
+    /**
+     * @test
+     */
+    public function multiplierHasToBeGreaterThan0(): void
```
Comment (OndraM) [inline tests/unit/Model/Command/BoostTest.php]:
```suggestion
    public function shouldNotAllowMultiplierLessThan0(): void
```

File: tests/unit/Model/Command/BoostTest.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $this->expectException(DomainException::class);
+        $this->expectExceptionMessage('Provided "-1" is not greater than "0".');
+        $boost = Boost::create('valid_to >= NOW()', -1);
```
Comment (OndraM) [inline tests/unit/Model/Command/BoostTest.php]:
```suggestion

        Boost::create('valid_to >= NOW()', -1);
```

File: README.md [hunk 5]
```diff
@@ -227,6 +227,15 @@ $recommendations = $response->getRecommendation()->getData();
 // }
 ```
 
+You can further modify which items will be reccomended by providing boosting rules. Priority of items matching the
```
Comment (OndraM) [inline README.md]:
```suggestion
You can further modify which items will be recommended by providing boosting rules. Priority of items matching the
```

File: README.md [hunk 6]
```diff
@@ -227,6 +227,15 @@ $recommendations = $response->getRecommendation()->getData();
...
+MQL `$criteria` will be multiplied by the value of `multiplier`:
+
+```php
+$reccomendation = UserRecommendation::create('user-id', 5, 'test-scenario', 1.0, 3600)
```
Comment (OndraM) [inline README.md]:
```suggestion
$recommendation = UserRecommendation::create('user-id', 5, 'test-scenario', 1.0, 3600)
```

File: README.md (hunk 5 above)
Comment (OndraM) [inline README.md]: I'd move this whole added section right before "From $response, you can also access rest of the data:".

Or maybe better merge it with the example following "You can also set more granular options of the recommendation command", something like this:

```php
$recommendation = UserRecommendation::create('user-id', 5, 'test-scenario', 1.0, 3600);
$recommendation->setFilters(['for_recommendation = 1'])
    ->addBoost(Boost::create('valid_to >= NOW()', 2)) // Priority of items matching the Booster criteria will be multiplied by the value of multiplier
    ->addBoost(Boost::create('for_recommendation = 1', 3.5))
    ->setMinimalRelevance(MinimalRelevance::HIGH())
    ->enableHardRotation();

$response = $matej->request()
    ->recommendation($recommendation)
    ->send();
```

**Review by OndraM — CHANGES_REQUESTED** [review]

**Comment by JakubTesarek** [comment]
All fixed. I'll squash on merge

File: README.md [hunk 7]
```diff
@@ -193,10 +193,12 @@ $recommendations = $response->getRecommendation()->getData();
 You can also set more granular options of the recommendation command:
 
 ```php
-$recommendation = UserRecommendation::create('user-id', 5, 'test-scenario', 1.0, 3600);
+$recommendation = UserRecommendation::create('user-id', 5, 'test-scenario', 1.0, 3600)
```
Comment (OndraM) [inline README.md]:
```suggestion
$recommendation = UserRecommendation::create('user-id', 5, 'test-scenario', 1.0, 3600);
```

**Review by OndraM — CHANGES_REQUESTED** [review]
I found just one glitch in readme, otherwise ✅ 

Thanks 👍

File: src/Model/Command/UserRecommendation.php (hunk 2 above)
Comment (JakubTesarek) [reply src/Model/Command/UserRecommendation.php]: fixed

**Review by OndraM — APPROVED** [review]

**Pull request merged**